cluster/cluster.go

```diff
@@ -0,0 +1,431 @@
+	_, err := psvc.AddInstance([]services.PlacementInstance{inst})
+	if err != nil {
+		return nil, err
+	}
+	c.usedInstances[inst.ID()] = inst
+	c.spares = c.spares[1:]
+	return inst, nil
+}
+
+func (c *m3dbCluster) RemoveInstance(i env.M3DBInstance) error {
+	if status := c.Status(); status != ClusterStatusRunning && status != ClusterStatusInitialized {
+		return errClusterUnableToAlterPlacement
+	}
+
+	if _, ok := c.usedInstances[i.ID()]; !ok {
+		return errInstanceNotInUse
+	}
+	psvc := c.placementSvc
+	_, err := psvc.RemoveInstance(i.ID())
+	if err != nil {
+		return err
+	}
+	delete(c.usedInstances, i.ID())
+	c.spares = append(c.spares, i)
+	return nil
+}
+
+func (c *m3dbCluster) ReplaceInstance(oldInstance env.M3DBInstance) (env.M3DBInstance, error) {
+	if status := c.Status(); status != ClusterStatusRunning && status != ClusterStatusInitialized {
+		return nil, errClusterUnableToAlterPlacement
+	}
+
+	if _, ok := c.usedInstances[oldInstance.ID()]; !ok {
+		return nil, errInstanceNotInUse
+	}
+
+	numSpares := len(c.spares)
+	if numSpares < 1 {
+		return nil, errInsufficientCapacity
+	}
+
+	spareInstance := c.spares[0]
+	psvc := c.placementSvc
+	_, err := psvc.ReplaceInstance(oldInstance.ID(), []services.PlacementInstance{spareInstance})
+	if err != nil {
+		return nil, err
+	}
+
+	delete(c.usedInstances, oldInstance.ID())
+	c.usedInstances[spareInstance.ID()] = spareInstance
+	c.spares = append(c.spares[1:], oldInstance)
+
+	return spareInstance, nil
+}
+
+func (c *m3dbCluster) Spares() []env.M3DBInstance {
+	return c.spares
+}
+
+func (c *m3dbCluster) Reset() error {
+	if status := c.Status(); status != ClusterStatusRunning && status != ClusterStatusInitialized &&
+		status != ClusterStatusSetup && status != ClusterStatusError {
+		return errClusterUnableToReset
+	}
+
+	var (
+		wg        sync.WaitGroup
+		lock      sync.Mutex
+		instances = c.env.Instances()
+		multiErr  xerrors.MultiError
+	)
+
+	// setup the instances, in parallel
+	for idx := range instances {
+		inst := instances[idx]
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			if err := inst.Reset(); err != nil {
+				lock.Lock()
+				multiErr = multiErr.Add(err)
+				lock.Unlock()
+				return
+			}
+
+			lock.Lock()
+			c.spares = append(c.spares, inst)
+			lock.Unlock()
+		}()
+	}
+	wg.Wait()
+
+	return c.markStatus(ClusterStatusSetup, multiErr.FinalError())
+}
+
+func (c *m3dbCluster) markStatus(status Status, err error) error {
+	c.statusLock.Lock()
+	defer c.statusLock.Unlock()
+
+	if err == nil {
+		c.status = status
+		return nil
+	}
+
+	c.status = ClusterStatusError
+	c.lastErr = err
+	return err
+}
+
+func (c *m3dbCluster) Teardown() error {
+	if status := c.Status(); status == ClusterStatusUninitialized {
+		return errClusterUnableToTeardown
+	}
+
+	var (
+		wg        sync.WaitGroup
+		lock      sync.Mutex
+		instances = c.env.Instances()
+		multiErr  xerrors.MultiError
+	)
+
+	// setup the instances, in parallel
+	for idx := range instances {
```

Comment:
this code seems to be repeated a lot, can make a generic one that takes in different func(inst) like Teardown() or Start() etc?